New filter schema

Akarys42 opened the pull request:
Implements the new filter related schema described in #479. The definition of each model can be found on this issue.

It adds a bunch of new routes under `/bot/filters` for each model, for example `/bot/filter/filter_lists` will provide GET, POST, PUT and DELETE for the `FilterList` model.

It also adds a migration (number 70) that will convert our current filters to the new schema and set some defaults that should mimic our current setup. Note that it will use a default value if the description isn't set, we will have to set one.

Tests have also been updated. I tried to make them modular in case we ever want to modify the design and to avoid having 800 lines of copy pasted code.

Most of this PR is just REST framework boilerplate, enjoy!

Closes #479

Akarys42 committed:
Filter: new schema

This commit adds new filter schema as described in #479

Akarys42 committed:
Filters: Add new models to Django Admin

Akarys42 committed:
Filters: Add migration to the new model

This will take the currently defined filter list and put them inside the new schema while trying to keep defaults similar to our current setup.

Akarys42 committed:
Filters: hook the new models into the REST API

Akarys42 committed:
Filters: update tests to the new schema

Akarys42 committed:
Filters: allowed_category -> allowed_categories

mbaruh commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,187 @@
+    """Choice between allow or deny for a list type."""
+
+    ALLOW: 1
+    DENY: 0
```
What is the motivation to make it integer choices, rather than text choices as the infraction types below?

Akarys42 replied:
It was just a matter of having a little bit more lightweight field, as this is roughly just a boolean field, there are only two states. I don’t mind changing it though.

mbaruh commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,187 @@
+
+
+# Valid special values in ping related fields
+VALID_PINGS = ("everyone", "here", "moderators", "onduty", "admins")
```
Is it necessary to define these? we could just have a list of role ID's.

Akarys42 replied:
There is no role ID for `here` though, how do we want to handle this?

mbaruh commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,187 @@
+    name = models.CharField(max_length=50, help_text="The unique name of this list.")
+    list_type = models.IntegerField(
+        choices=FilterListType.choices,
+        help_text="Whenever this list is an allowlist or denylist"
```
```suggestion
        help_text="Whether this list is an allowlist or a denylist"
```

mbaruh commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,187 @@
+        validators=(validate_ping_field,),
+        help_text="Who to ping when this filter triggers."
+    )
+    filter_dm = models.BooleanField(help_text="Whenever DMs should be filtered.")
```
```suggestion
    filter_dm = models.BooleanField(help_text="Whether DMs should be filtered.")
```

mbaruh commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,187 @@
+        help_text="Who to ping when this filter triggers on a DM."
+    )
+    delete_messages = models.BooleanField(
+        help_text="Whenever this filter should delete messages triggering it."
```
```suggestion
        help_text="Whether this filter should delete messages triggering it."
```

mbaruh commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,187 @@
+    )
+    bypass_roles = ArrayField(
+        models.BigIntegerField(),
+        help_text="Roles and users who can bypass this filter."
```
How do you tell whether an ID belongs to a role or to a user? should be limited to roles IMO.

Akarys42 replied:
We usually do that by trying to resolve it as a role and then a user. We cannot run into conflicts this way.

mbaruh commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,187 @@
+        help_text="Roles and users who can bypass this filter."
+    )
+    enabled = models.BooleanField(
+        help_text="Whenever ths filter is currently enabled."
```
```suggestion
        help_text="Whether this filter is currently enabled."
```

mbaruh commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,187 @@
+    default_action = models.ForeignKey(
+        "FilterAction",
+        models.CASCADE,
+        help_text="The default action to perform."
```
Not really clear what that means
```suggestion
        help_text="What action to perform on the triggering user."
```

mbaruh commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,187 @@
+    default_range = models.ForeignKey(
+        "ChannelRange",
+        models.CASCADE,
+        help_text="Where does this filter apply."
```
```suggestion
        help_text="The channels and categories in which this filter applies."
```

mbaruh commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,187 @@
+        return f"Filter {self.content!r}"
+
+
+class FilterOverride(models.Model):
```
Can we call it SettingsOverride? since we're overriding the settings.

mbaruh commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,187 @@
+    bypass_roles = ArrayField(models.IntegerField(), null=True)
+    enabled = models.BooleanField(null=True)
+    filter_action = models.ForeignKey("FilterAction", models.CASCADE, null=True)
+    filter_range = models.ForeignKey("ChannelRange", models.CASCADE, null=True)
```
I think these names should also be the names in FilterSettings.

mbaruh reviewed (changes requested):
Thank you, that was record speed.

I think it's a good start so that we have something to test with. Aside from the minor comments below, there are two things I think are important:
* An endpoint to recursively fetch all data in the schema, so that we don't have to join tables manually on the bot side, and just make one API request on startup.
* Very detailed docs, to the point of saying where exactly you need to go to remove a kind of setting.

MarkKoz commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,187 @@
+        help_text="Whenever this list is an allowlist or denylist"
+    )
+
+    filters = models.ManyToManyField("Filter", help_text="The content of this list.", default=[])
```
This should not be a many-to-many relationship. It doesn't make sense for a filter to belong to multiple lists. It should be a one-to-many relationship, which is simply defined with a foreign key. Django supports [following relationship backwards](https://docs.djangoproject.com/en/3.2/topics/db/queries/#following-relationships-backward), if needed.

MarkKoz commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,187 @@
+        )
+
+    def __str__(self) -> str:
+        return f"Filter {'allow' if self.list_type == 1 else 'deny'}list {self.name!r}"
```
Don't hard-code this. It's like a normal Python enum, so you can do `FilterListType(self.list_type)` to get the name from the value. If it had labels defined, then, of course, you could use `FilterListType(self.list_type).label`.

MarkKoz commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,187 @@
+class FilterAction(models.Model):
+    """The action to take when a filter is triggered."""
+
+    user_dm = models.CharField(
```
`dm_content` is a better name.

MarkKoz commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,187 @@
+    infraction_duration = models.DurationField(
+        null=True,
+        help_text="The duration of the infraction. Null if permanent."
+    )
```
There should be some validation (*ideally at the database level*) for these. It would make the API more robust. It is likely a mistake if a duration or reason is set but a type isn't. Furthermore, some infraction types don't support permanent durations. In fact, I'm not sure if this is normalised given that there's (sort of?) a relationship between the type and duration.

mbaruh replied:
There isn't actually anything preventing a mute from being permanent at the database level, we simply chose to not expose such a command on the bot side, but it's still possible to edit a mute to become permanent.
Likewise, notes and warnings seem to be marked as permanent, at least on the bot side.

MarkKoz replied:
I should have said it the other way around. Some infractions aren't meant to expire, like notes and warnings. That's a semantic limitation, not a technical one. However, there is validation imposed that prevents it.

We could allow it here and defer the error to when it actually tries to apply an infraction, but that is a bad idea.

MarkKoz commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,187 @@
+    """The action to take when a filter is triggered."""
+
+    user_dm = models.CharField(
+        max_length=1000,
```
How did you decide this number for the dm and the reason?

mbaruh replied:
For infraction reasons I believe there are issues with pagination on the bot side when they're too long. I know we limited nomination reasons to 1000 characters, and recall we did something similar with infractions.

For DM's however the limit is probably not 1000.

MarkKoz replied:
Can this import the infraction model and use its reason limit here instead of having two separate constants that need to be maintained in tandem?

mbaruh replied:
This seems like a good approach, yeah.

MarkKoz commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,187 @@
+
+    content = models.CharField(max_length=100, help_text="The definition of this filter.")
+    description = models.CharField(max_length=200, help_text="Why this filter has been added.")
+    additional_field = models.BooleanField(null=True, help_text="Implementation specific field.")
```
This is really limited, both in count and in type. It's conceivable for filters to use a value other than a boolean for setting and to use more than one filter-specific setting.

MarkKoz commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,187 @@
+        models.SET_NULL,
+        null=True,
+        help_text="Override the default settings."
+    )
```
"Override" is ambiguous. What is it overriding? Only the help text would be able to explain that. Please name this "settings".

MarkKoz commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,187 @@
+    disallowed_categories = ArrayField(models.IntegerField())
+    allowed_channels = ArrayField(models.IntegerField())
+    allowed_categories = ArrayField(models.IntegerField())
+    default = models.BooleanField()
```
What does this mean?

mbaruh replied:
I think "default" should be removed.

MarkKoz commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,187 @@
+    disallowed_channels = ArrayField(models.IntegerField())
+    disallowed_categories = ArrayField(models.IntegerField())
+    allowed_channels = ArrayField(models.IntegerField())
+    allowed_categories = ArrayField(models.IntegerField())
```
Something to keep in mind:

Arrays may be an obvious example of a non-atomic type. However, 1NF is ambiguous when it comes to atomicity. Arrays _can_ be treated as atomic types, but I believe this means one must not intend to establish any relationships with the individual elements in the array. My further understanding is that as a consequence, one should not intend to perform queries like "which filters does channel x have?"

MarkKoz replied:
There should be validation (again, ideally at the database level) that

1. Each pair of arrays is mutually exclusive.
2. A duplicate value is not inserted into an array.

Starting to make be doubt if arrays are a suitable data type given such constraints. The four arrays could instead be 2 tables that have 2 columns - one for the ID and another for whether its allowed/disallowed. A primary key on the ID would address both of the validation points above.

mbaruh replied:
Is it acceptable to have a field to tell whether it's a channel or a category? I'd rather these weren't split into multiple tables, as then it complicates grouping them on the bot side.

MarkKoz replied:
Yes, I don't see why it wouldn't be acceptable. I don't believe it's possible for IDs to overlap either, if that was a concern.

As for your other point, keep in mind DRF lets you serialise the data in any way you want. It is not limited to being a 1:1 representation of the DB schema. Granted, it's easier when they do align, so the added work and any performance impacts should be weighed in along with  the benefit I mentioned in the original comment.

kosayoda replied:
Given the situation where a channel/category could be whitelisted for a filter and blacklisted for another, how would a primary key on the ID work to address the validation issues? From what I can see of the current implementation, a `ChannelRange` entry is just the domain in which a filter applies.

MarkKoz replied:
It's been a long time since I've looked into this. I think it'd have to be a composite PK on the channel ID and the filter ID.

MarkKoz commented on pydis_site/apps/api/models/bot/filters.py:
```diff
@@ -0,0 +1,187 @@
+        if value.isnumeric():
+            continue
+
+        raise ValidationError(f"{value!r} isn't a valid ping type.")
```
I propose to split the columns into roles and users so that each can contain homogenous types. Allowing arbitrary IDs for roles offers more flexibility and gets rid of some of the unintuitiveness and complexity of mixing roles and users in a single column.

The alternative to 2 arrays would be a single array with two columns - one for the ID and another for the type.

mbaruh replied:
I prefer the latter.

MarkKoz replied:
Sorrry, I made a mistake. To be clear, I meant a single table with 2 columns, not a single array.

MarkKoz reviewed (changes requested):
I'm only reviewing the schema for now.

A mistake you've made is conflating one-to-many relationships with one-to-one relationships. Recall that a foreign key describes the former, not the latter. However, the relationships between settings and filters is one-to-one, as well as the relationship between settings and actions/ranges.

One-to-one relationships are kind of strange to me in the world of database design, since they _can_ be merged into a single table. There are some technical reasons for separating them (performance, different security policies, etc.), but I think the primary reason it's done here is for semantics. The argument is that it betters the understanding of the schema, and aides in the organisation and maintenance of it. 

The distinction between some of the models in the one-to-one-relationships seem a bit weak. Furthermore, it's not consistent e.g. the infraction fields arguably represent an infraction, so they should be in a separate table for semantic purposes too.

However, at least for the filters to settings relationship, I feel the decision to split them can still hold up. There would be a stronger case if the override was the exact same schema as the defaults (in which case the tables could be merged too), but unfortunately that seems impossible due to one needing to have nullable fields.

The idiomatic way to implement one-to-one relationships in Django is through [inheritance](https://docs.djangoproject.com/en/3.2/topics/db/models/#model-inheritance). The question is whether each model should have its own table or if the fields of the abstract parents should be instead in the child tables.

The latter sounds better to me, since I believe it'd avoid cross-table joins. Since the schema is managed through Django, it will still give the illusion of being separate where it matters. I wonder if it's possible to let the child dictate whether the inherited fields should be nullable; this would avoid defining two separate settings classes.

Note that, if needed, serialisation can be customised to organise things in a different manner than the schema represents it.

kosayoda committed:
Improve help text message.

kosayoda committed:
Fix faulty model enumeration.

This also allows us to simplify the str dunder for a FilterList.

kosayoda committed:
Improve name of dm sent to triggered user.

kosayoda committed:
Migrate misc field names and help text changes.

kosayoda committed:
Correct Filter-FilterList relationship.

Instead of a many-many relationship, one filterlist has multiple
filters. Nested serialization is read-only by default, so not
all CRUD methods are implemented yet for the FilterList viewset.

ks129 committed:
Remove one-to-one relationships from filters tables

ks129 committed:
Remove old one-to-one filters relationships serializers, views and URLs

ks129 committed:
Add AbstractModelMeta mixin

ks129 committed:
Update filters migrations

ks129 committed:
Remove default field from FilterSettingsMixin and migration

ks129 committed:
Remove old models from migration

ks129 committed:
Update filters API to actually work

ks129 committed:
Add basic validation for infraction fields + use common infraction types

ks129 committed:
Add validation to filters to not allow duplicated channels and categories

ks129 committed:
Add validation to filters to not allow duplicates + additional_field -> JSON

ks129 committed:
Move filters validations to serializers

ks129 committed:
Add merge migration

ks129 committed:
Fix linting

D0rs4n committed:
Adjust Filter JSON Schema

From now on the Serializer will have a different JSON representation than the
table schema itself, conforming to the format needed on the bot-side.

D0rs4n committed:
Adjust FilterList Representation

From now on the FilterList Serializer will contain a settings field with all the settings that were listed previously, on the model.

D0rs4n committed:
Adjust FilterList Schema to group settings into subcategories

- This commit patches the FilterList serializer's schema, and puts the settings into the relevant subcategories.

D0rs4n committed:
Patch Filter and FilterList Serializer validation logic and representation

- This commit patches an error with the FilterListSerializer validation logic, so that it won't raise an error when an optional field is not present.
- It also adds the `enabled` and `delete_messages` fields, to the FilterSerializer's representation
- Furthermore the commit introduces minor bug patches, regarding DRF Serializer Fields.

D0rs4n committed:
Prepare FilterList and Filter models, serializers for the new filter schema

- Rename channel scope fields:
 - "allowed" -> "disabled"
eg.: "allowed_channels" -> "disabled_channels"
- Rename FilterLists` names:
filter_token -> tokens
domain_name -> domains
guild_invite -> invites
file_format -> formats
- Patch the docs and validators accordingly.

D0rs4n committed:
Patch Filter/FilterList's default values and add new fields

- Patch default values, so that further implementations can be performed on the bot side
- Add three new fields: "send_alert", and in settings under the "server_message" field: "send_message_text", and "server_message_embed" fields.
- Patch documentation, and validators accordingly.
- Perform further patches, and minor corrections.

D0rs4n committed:
Include 'dm_content ' field under Infraction settings in Filters/FilterLists

D0rs4n committed:
Correct 'Redirect' FilterLists' default values.

mbaruh committed:
Merged infraction and notification settings in JSON

The settings for infracting and notifying the user were merged under one field, which is renamed to "infraction_and_notification".
The only place which sends a message in the server by default is the antimalware, the rest try to DM the user first, and antimalware can do the same.

This avoids complications which may result from the filtering cog trying to send two messages: one for the defined server message, and another for a failed DM.

mbaruh committed:
Merge branch 'main' into new-filter-schema

mbaruh committed:
Merge migrations and correct filter defaults to be all null

mbaruh committed:
Allow filter descriptions to be null

mbaruh committed:
Use singular nouns for filter list names

D0rs4n committed:
Patch a minor issue with FilterList field naming in migrations

mbaruh committed:
Changed channeld fields to char arrays, merged migrations

- The fields concerning channels were changed to contains strings instead of integers in order to allow specifying channels and categories by name.

The migrations were merged into a single migration.

mbaruh committed:
Extensions list is ALLOW, not DENY

mbaruh committed:
Create placeholder value for dm embed content in ext list

Some value is needed to signal the bot a message should be sent for a blocked extension. The value itself will be changed at runtime, but this allows avoiding the bot code delving into the exact API response format.

mbaruh committed:
Add invites denylist to the migration

mbaruh committed:
Refine DM content for domains

mbaruh committed:
Change ping fields names

mbaruh committed:
Remove role validation

Roles can be either IDs or names, so the current validation is not relevant anymore. Furthermore the ping fields can accept user IDs or names.

D0rs4n committed:
Sync Filter models with relating migrations, adjust code consistency

D0rs4n committed:
Add UniqueConstraint to the Filter model

- The UniqueConstraint includes every field, except for id and description.
